src/neptune/new/internal/disk_queue.py

```diff
@@ -111,32 +122,40 @@ def _skip_and_get(self) -> Tuple[Optional[T], int]:
                         ack_version,
                         ver,
                     )
-                return obj, ver
+                return serialized
 
-    def _get(self) -> Tuple[Optional[T], int]:
+    def _get_serialized(self) -> Optional[dict]:
         _json = self._reader.get()
         if not _json:
             if self._read_file_version >= self._write_file_version:
-                return None, -1
+                return None
             self._reader.close()
             self._read_file_version = self._next_log_file_version(self._read_file_version)
             self._reader = JsonFileSplitter(self._get_log_file(self._read_file_version))
             # It is safe. Max recursion level is 2.
-            return self._get()
-        try:
-            return self._deserialize(_json)
-        except Exception as e:
-            raise MalformedOperation from e
+            return self._get_serialized()
+        return _json
 
     def get_batch(self, size: int) -> Tuple[List[T], int]:
-        first, ver = self.get()
-        if not first:
-            return [], ver
+        if self._should_skip_to_ack:
+            serialized_first = self._skip_and_get_serialized()
+        else:
+            serialized_first = self._get_serialized()
+        if not serialized_first:
+            return [], -1
+
+        cur_batch_size = len(serialized_first)
+        first, ver = self._deserialize(serialized_first)
         ret = [first]
         for _ in range(0, size - 1):
-            obj, next_ver = self._get()
-            if not obj:
+            if cur_batch_size >= self._max_batch_size_bytes:
+                break
+            serialized_obj = self._get_serialized()
+            if not serialized_obj:
                 break
+            cur_batch_size += len(serialized_obj)
```

Comment:
`_get_serialized()` returns `dict`. So unfortunately `len(serialized_obj)` is number of keys in `dict`, not size of json in bytes, isn't it? Maybe we should add returning size to `JsonFileSplitter.get()`, what do you think?

Reply:
I think it's a good idea. I've added a get_with_size function to the JsonFileSplitter, as well as a unit test for it. I had to refactor the DiskQueue as well.